vdj_ann/src/transcript.rs

```diff
@@ -7,187 +7,252 @@ use crate::refx::RefData;
+struct Annotation {
+    tig_start: usize,
+    match_length: usize,
+    ref_id: usize,
+    ref_start: usize,
+}
+
+fn find_inframe_vdj_pair(vstarts: Vec<Vstart>, jstops: Vec<Jstop>) -> Option<(Vstart, Jstop)> {
+    let mut vj_combinations: Vec<(Vstart, Jstop, i32)> = iproduct!(vstarts, jstops)
+        .map(|(v, j)| (v, j, j.tig_stop as i32 - v.tig_start as i32))
+        .filter(|(_, _, n)| n > &0)
+        .filter(|(_, _, n)| n % 3 == 1)
+        .collect();
+    vj_combinations.sort_by_key(|x| x.2);
+    let inframe_pair = vj_combinations.last().map(|(v, j, _)| (*v, *j));
+    inframe_pair
+}
+
+fn evaluate_contig_status(
+    vdj_chain: VdjChain,
+    ann: &[(i32, i32, i32, i32, i32)],
+    reference: &RefData,
+    contig: &DnaString,
+) -> Option<ContigStatus> {
+    let vj_pair = VdjChainSpecificVJGenes::from(vdj_chain);
+    let rheaders = &reference.rheaders;
+    let refs = &reference.refs;
+    let annotation: Vec<Annotation> = ann
+        .iter()
+        .map(
+            |(tig_start, match_length, ref_id, ref_start, _)| Annotation {
+                tig_start: *tig_start as usize,
+                match_length: *match_length as usize,
+                ref_id: *ref_id as usize,
+                ref_start: *ref_start as usize,
+            },
+        )
+        .collect();
+    let mut vstarts: Vec<Vstart> = annotation
+        .iter()
+        .filter(|a| reference.is_v(a.ref_id))
+        .filter(|a| rheaders[a.ref_id].contains(vj_pair.v_type.as_str()))
+        .filter(|a| a.ref_start == 0)
+        .map(|a| Vstart {
+            ref_id: a.ref_id,
+            tig_start: a.tig_start,
+        })
+        .collect();
+    let jstops: Vec<Jstop> = annotation
+        .iter()
+        .filter(|a| reference.is_j(a.ref_id))
+        .filter(|a| rheaders[a.ref_id].contains(vj_pair.j_type.as_str()))
+        .filter(|a| a.ref_start + a.match_length == refs[a.ref_id].len())
+        .map(|a| Jstop {
+            ref_id: a.ref_id,
+            tig_stop: a.tig_start + a.match_length,
+        })
+        .collect();
+
+    if vstarts.is_empty() & jstops.is_empty() {
+        return None;
+    }
+
+    let mut contig_status = ContigStatus {
+        full_length: Some(!vstarts.is_empty() & !jstops.is_empty()),
+        ..Default::default()
+    };
+
+    // filter vstarts to require START codon
+    vstarts.retain(|v| have_start(contig, v.tig_start));
+    contig_status.has_v_start = match (contig_status.full_length, vstarts.is_empty()) {
+        (Some(true), false) => Some(true),
+        (_, _) => Some(false),
+    };
+
+    let inframe_pair = find_inframe_vdj_pair(vstarts, jstops);
+    contig_status.in_frame = Some(inframe_pair.is_some());
+
+    contig_status.no_premature_stop = if let Some((vstart, jstop)) = inframe_pair {
+        Some(
+            !(vstart.tig_start..jstop.tig_stop - 3)
+                .step_by(3)
+                .any(|j| have_stop(contig, j)),
+        )
+    } else {
+        None
+    };
```

Comment:
Since `contig_status.no_premature_stop` is already `None`, you can invert the assignment. Similar refactors for related cases nearby.
```suggestion
    if let Some((vstart, jstop)) = inframe_pair {
        contig_status.no_premature_stop = Some(
            !(vstart.tig_start..jstop.tig_stop - 3)
                .step_by(3)
                .any(|j| have_stop(contig, j)),
        )
    };
```